### PR TITLE
Add Component Property API

### LINKED ISSUE #2715: Add Properties to components API
### Current Behavior

Currently you can add properties to projects but you can't for components.

### Proposed Behavior

Replicate the project properties API for components.  This would allow my team to make sure of DT in our FLOSS approval process by including properties about the workflow state and other information. 

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

### PR DESCRIPTION
**Pull request by rkesters**
### Description

Added the same Property API to Components that exists for Projects as requested in  #2715 

### Addressed Issue
closes #2715 

### Additional Details

* Did not update the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs)  as Project Property API did not appear to be documented.

* Updated [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade) by adding a marking updater to increase the schema version.   If I did this wrong please let me know what I should have done. 

### Checklist

- [X] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- [ ] This PR fixes a defect, and I have provided tests to verify that the fix is effective
- [X] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [X] This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)
- [X] This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly

### CODE REVIEW

File: src/main/java/org/dependencytrack/persistence/QueryManager.java [hunk 1]
```diff
@@ -488,6 +489,23 @@ public PaginatedResult getComponents(final boolean includeMetrics) {
...
+    }
+
+	public  ComponentProperty getComponentProperty(final Component component
+	, final String groupName, final String propertyName) {
+        return getComponentQueryManager().getComponentProperty(component, groupName, propertyName);
+    }
+
+	public ComponentProperty createComponentProperty(final Component component, 
+	final String groupName, final String propertyName,
+	final String propertyValue, final ComponentProperty.PropertyType propertyType,
```
[the remark this answers is not in the document]
Comment (rkesters) [reply src/main/java/org/dependencytrack/persistence/QueryManager.java]: made recommended change

File: src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                    } else {
+                        return Response.status(Response.Status.NOT_FOUND).entity("The project property could not be found.").build();
+                    }
+                } else {
+                    return Response.status(Response.Status.FORBIDDEN).entity("Access to the specified project is forbidden").build();
+                }
+            } else {
+                return Response.status(Response.Status.NOT_FOUND).entity("The project could not be found.").build();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rkesters) [reply src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java:241]: other instances of this issue currently exist.

@sonatype-lift ignore

File: src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java [hunk 3]
```diff
@@ -105,6 +106,54 @@ public PaginatedResult getComponents() {
...
+     * @param component the Component to create the property for
+     * @param groupName the group name of the property
+     * @param propertyName the name of the property
+     * @param propertyValue the value of the property
+     * @param propertyType the type of property
+     * @param description a description of the property
+     * @return the created ComponentProperty object
+     */
+    public ComponentProperty createComponentProperty(final Component component, final String groupName, final String propertyName,
+                                                 final String propertyValue, final ComponentProperty.PropertyType propertyType,
```
[the remark this answers is not in the document]
Comment (rkesters) [reply src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java]: fixed

File: cp.txt
Comment (nscuro) [inline cp.txt:1]: Seems to be a leftover from testing, please remove this file.

File: src/main/java/org/dependencytrack/model/ComponentProperty.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.Size;
+import java.io.Serializable;
+
+/**
+ * User-defined key/value model for individual projects.
+ *
+ * @author Steve Springett
+ * @since 3.0.0
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/model/ComponentProperty.java:43]: Nit: Needs to be updated.

```suggestion
 * User-defined key/value model for individual components.
 *
 * @author Steve Springett
 * @since 4.9.0
```

File: src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java
Comment (nscuro) [inline src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java:1]: Please revert the unrelated code style changes. We appreciate the intention, but it's causing too much noise 😅

File: src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import io.swagger.annotations.Authorization;
+
+/**
+ * JAX-RS resources for processing ProjectProperties
+ *
+ * @author Steve Springett
+ * @since 3.4.0
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java:54]: Nit: Needs to be updated.

```suggestion
 * JAX-RS resources for processing ComponentProperties
 *
 * @author Steve Springett
 * @since 4.9.0
```

File: src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                return Response.status(Response.Status.NOT_FOUND).entity("The component could not be found.").build();
+            }
+        }
+    }
+
+    @PUT
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "Creates a new project property",
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java:107]:
```suggestion
            value = "Creates a new component property",
```

File: src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "Updates a project property",
+            response = ComponentProperty.class
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 401, message = "Unauthorized"),
+            @ApiResponse(code = 403, message = "Access to the specified project is forbidden"),
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java:169]:
```suggestion
            @ApiResponse(code = 403, message = "Access to the specified component is forbidden"),
```

File: src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+                        qm.close();
+                        if (ComponentProperty.PropertyType.ENCRYPTEDSTRING == property.getPropertyType()) {
+                            property.setPropertyValue(ENCRYPTED_PLACEHOLDER);
+                        }
+                        return Response.status(Response.Status.CREATED).entity(property).build();
+                    } else {
+                        return Response.status(Response.Status.CONFLICT).entity("A property with the specified component/group/name combination already exists.").build();
+                    }
+                } else {
+                    return Response.status(Response.Status.FORBIDDEN).entity("Access to the specified project is forbidden").build();
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java:152]:
```suggestion
                   return Response.status(Response.Status.FORBIDDEN).entity("Access to the specified component is forbidden").build();
```

File: src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "Returns a list of all ComponentProperties for the specified component",
+            response = ComponentProperty.class,
+            responseContainer = "List"
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 401, message = "Unauthorized"),
+            @ApiResponse(code = 403, message = "Access to the specified project is forbidden"),
+            @ApiResponse(code = 404, message = "The project could not be found")
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/ComponentPropertyResource.java:72]:
```suggestion
            @ApiResponse(code = 404, message = "The component could not be found")
```

File: src/main/java/org/dependencytrack/upgrade/v490/v490Updater.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ */
+package org.dependencytrack.upgrade.v490;
+
+import alpine.common.logging.Logger;
+import alpine.persistence.AlpineQueryManager;
+import alpine.server.upgrade.AbstractUpgradeItem;
+
+import java.sql.Connection;
+
+public class v490Updater extends AbstractUpgradeItem {
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/upgrade/v490/v490Updater.java:27]: We will not need an `UpgradeItem` for this change. The underlying ORM will take care of creating the new table automatically.

**Review by nscuro — CHANGES_REQUESTED** [review]
Thanks for the PR, @rkesters! Few minor changes required.

**Comment by melba-lopez** [comment]
@rkesters thanks for making these enhancements!  Are you able to make the changes requested by @nscuro?  Please make sure to update your branch as well 😄

File: cp.txt
Comment (rkesters) [reply cp.txt:1]: REMOVED

File: src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java
Comment (rkesters) [reply src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java:1]: I believe I corrected the items you are referring to.

**Comment by rkesters** [comment]
> @rkesters thanks for making these enhancements! Are you able to make the changes requested by @nscuro? Please make sure to update your branch as well 😄

@melba-lopez I have responded to the comments and rebased the branch.
